rpc: getblockfrompeer

Sjors opened the pull request:
This adds an RPC method to fetch a block directly from a peer. This can used to fetch stale blocks with lower proof of work that are normally ignored by the node (`headers-only` in `getchaintips`).

Usage:
```
bitcoin-cli getblockfrompeer HASH peer_n
```

Closes #20155

Limitations:
* you have to specify which peer to fetch the block from
* the node must already have the header

Fi3 commented:
@Sjors Please note that I'm very new to C++ and I'm here only to learn.

Concept AKN, IMO can also be useful if the user need info about a pruned block.

Not sure if `getblockfrompeer` should persist the block, maybe just print it. For consistency the interface of `getblock` (verbosity 0, 1, 2) could be reused.
Maybe `getblockfrompeer` could just be a flag of `getblock`.

Also, IMO `GetBlockChecked` should notify the user to use `getblockfrompeer` for block that are not on disk.

**Check if the node is connected**
Checking for connectivity could be done with `connman->GetNodeCount` as in [mining.cpp](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/mining.cpp#L666).

**Mark the block as in flight**
Instead of mark the block as in flight and persist the block, I would add 2 fields to the public interface of `CNode` the first to notify `ProcessMessage` that a block with a certain hash is  expected, the second to check (in a loop) for the requested block.

Sjors commented:
@Fi3 being able to (temporarily) retrieve pruned blocks is useful too indeed.

It can be difficult to find nodes that still have a stale block, so I'd rather hold on to it. That also allows inspection using the regular methods, `getblock`, `getchaintips` and `getrawtransaction` (e.g. to learn about a specific doublespend). Pruned nodes will toss it automatically after a while (IIUC).

I also want to be able to fully validate the block. It's very time consuming, but it works: call `invalidateblock` on the main chain block for the same height, so reorgs to the block you just downloaded and verified it. Then call `reconsiderblock` to sync to the tip again.

BitMEXResearch commented:
Concept ACK

This would be useful for https://forkmonitor.info/ to assess the validity of shorter chains

benthecarman commented:
Concept ACK

This would be very useful for testing

Fi3 commented:
@Sjors 

> It can be difficult to find nodes that still have a stale block, so I'd rather hold on to it. 

I agree, didn't consider it.

luke-jr commented:
Concept -1

I think it would be better to just do the right thing if you `getblock` a block you don't have... Don't make users figure out which peer has it...?

Sjors commented:
@luke-jr my thinking was to expand it later, to try all peers: `getblockfrompeer HASH *`. Overloading `getblock` in a way that could make it unresponsive would seem controversial. Adding a `fetch_from_peer` argument to `getblock` could make sense as an alternative.

In practice none of the existing peers might have it, so you'll have to:
* manually connect to new peers and retry; or
* add an argument with an array of IPs to try (but not keep as a peer); or
* go through peers.dat entries

This could get complicated, so a fresh RPC seemed like a more future proof approach.

hamedtayeban commented:
0xC46086A8e6FD0013c9Fb4a5937fe3E75a7F957C6

hamedtayeban commented:
https://freewallet.org/id/5d88cebc/btc

Sjors commented:
I moved `FetchBlock` to the PeerManager while rebasing on #19910. Not sure if that's the right home for it, cc @jnewbery.

prusnak commented:
Does this work with the pruned node too? If yes, then this implements option 2) of https://github.com/bitcoin/bitcoin/issues/19312

Sjors commented:
It does, however I'm not sure how quickly the node re-prunes it.

Sjors commented:
Rebased. No longer needs to haul the mempool around.

Sjors commented:
Rebased after #21719

Sjors commented:
Linter:
> A new circular dependency in the form of "rpc/blockchain -> rpc/net -> rpc/blockchain" appears to have been introduced.

@MarcoFalke any thoughts on how to avoid that? I could move this new RPC method to `rpc/net`, or maybe these helpers should be in their own file?

MarcoFalke commented:
I guess by moving the `Ensure*` helpers to `rpc/util` (or a new module)?

jnewbery commented on src/net_processing.cpp:
```diff
@@ -246,6 +246,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Implement PeerManager */
     void CheckForStaleTipAndEvictPeers() override;
+    bool FetchBlock(const NodeId nodeid, const CBlockIndex* pindex) override;
```
No need to make the pass-by-value `NodeId nodeid` const.

Style guide is to use snake_case and not hungarian notation for parameter names:

```suggestion
    bool FetchBlock(const NodeId id, const CBlockIndex* index) override;
```

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
            (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, m_chainparams.GetConsensus()) < STALE_RELAY_AGE_LIMIT);
 }
 
+bool PeerManagerImpl::FetchBlock(const NodeId nodeid, const CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
I don't think this `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` is doing anything. This function is being called without `cs_main` (by `getblockfrompeer()`).

I think it's better to remove this annotation and lock `cs_main` within this function.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+{
+    PeerRef peer = GetPeerRef(nodeid);
+    if (peer == nullptr) return false;
+    uint32_t nFetchFlags = 0;
```
Move this down to where it's first used, and make const:

```suggestion
    const uint32_t nFetchFlags = state->fHaveWitness ? MSG_WITNESS_FLAG : 0;
```

Sjors replied:
I followed your suggestion below to just ignore pre-segwit nodes, so no need to have `nFetchFlags` around...

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (state->fHaveWitness) {
+        nFetchFlags |= MSG_WITNESS_FLAG;
+    }
+    std::vector<CInv> vInv(1);
+    vInv[0] = CInv(MSG_BLOCK | nFetchFlags, pindex->GetBlockHash());
```
Use initializer list ctor:

```suggestion
    std::vector<CInv> vInv{CInv(MSG_BLOCK | nFetchFlags, pindex->GetBlockHash())};
```

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
 }
 
+bool PeerManagerImpl::FetchBlock(const NodeId nodeid, const CBlockIndex* pindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
```
Perhaps return false if we're importing or reindexing (since the received `block` message will be ignored)

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (pindex->nHeight > node_sync_height) {
+        return false;
+    }
+    if (state->fHaveWitness) {
```
Perhaps just return false if the peer doesn't have `MSG_WITNESS_FLAG`. It's 2021. Receiving a block without witnesses isn't going to be useful.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+        this->m_connman.PushMessage(pnode, msgMaker.Make(NetMsgType::GETDATA, vInv));
+        return true;
+    });
+    MarkBlockAsInFlight(nodeid, pindex->GetBlockHash(), pindex);
```
Perhaps move this above the `ForNode()` call and don't send the `getdata` if it returns false, since that indicates that the block is already in flight. We don't want a faulty RPC client causing us to spam a peer with multiple requests for the same block.

Sjors replied:
Not sure if I got this right now...

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    }
+    std::vector<CInv> vInv(1);
+    vInv[0] = CInv(MSG_BLOCK | nFetchFlags, pindex->GetBlockHash());
+    bool success = m_connman.ForNode(nodeid, [this, vInv](CNode* pnode) {
```
Maybe add a log if the message is sent.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,33 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (state->fHaveWitness) {
+        nFetchFlags |= MSG_WITNESS_FLAG;
+    }
+    std::vector<CInv> vInv(1);
```
style guide:

```suggestion
    std::vector<CInv> invs(1);
```

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +767,46 @@ static RPCHelpMan getmempoolentry()
+    const CBlockIndex* pblockindex;
+    {
+        LOCK(cs_main);
+        pblockindex = g_chainman.m_blockman.LookupBlockIndex(hash);
```
Don't use the global chainman - use `EnsureChainman()`

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +767,46 @@ static RPCHelpMan getmempoolentry()
+    }
+
+    const UniValue &id_arg = request.params[1];
+    const NodeId nodeid = (NodeId) id_arg.get_int64();
+    const NodeContext& node = EnsureAnyNodeContext(request.context);
+    PeerManager& peerman = EnsurePeerman(node);
```
Move these to the top of the function - if these assumptions fail we should throw immediately instead of taking cs_main and looking up the block.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +767,46 @@ static RPCHelpMan getmempoolentry()
+{
+    uint256 hash(ParseHashV(request.params[0], "hash"));
+
+    const CBlockIndex* pblockindex;
+    {
+        LOCK(cs_main);
+        pblockindex = g_chainman.m_blockman.LookupBlockIndex(hash);
+    }
```
```suggestion
    const CBlockIndex* const pblockindex = WITH_LOCK(cs_main, return g_chainman.m_blockman.LookupBlockIndex(hash););
```

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +767,46 @@ static RPCHelpMan getmempoolentry()
+                    {"blockhash", RPCArg::Type::STR_HEX, RPCArg::Optional::NO, "The block hash"},
+                    {"nodeid", RPCArg::Type::NUM, RPCArg::Optional::NO, "The node ID (see getpeerinfo for node IDs)"},
+                },
+                RPCResult{RPCResult::Type::NONE, "", ""},
```
Return an empty object in case we want to extend the API in the future.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +767,46 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getblockfrompeer()
+{
+    return RPCHelpMan{"getblockfrompeer",
+                "\nAttempts to fetch block from a given peer.\n",
```
Indicative:

```suggestion
                "\nAttempt to fetch block from a given peer.\n",
```

jnewbery commented:
> I moved FetchBlock to the PeerManager while rebasing on #19910. Not sure if that's the right home for it, cc @jnewbery.

It's exactly the right place for it. `PeerManager` is the node's interface to net_processing.

Sjors committed:
rpc: move Ensure* helpers to server_util.h

Sjors commented:
Thanks for the feedback! I moved the Ensure* helpers to `util.h` and addressed the above comments.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -764,6 +716,43 @@ static RPCHelpMan getmempoolentry()
+                },
+        [&](const RPCHelpMan& self, const JSONRPCRequest& request) -> UniValue
+{
+    const UniValue &id_arg = request.params[1];
+    const NodeId nodeid = (NodeId) id_arg.get_int64();
```
Perhaps join these two lines. No point in having a `id_arg` var that just gets used once and thrown away:

```suggestion
    const NodeId nodeid = static_cast<NodeId>(request.params[1].get_int64());
```

jnewbery reviewed (commented):
Looking good (apart from the linker failure :slightly_smiling_face: )

jnewbery commented on src/rpc/util.cpp:
```diff
@@ -20,6 +25,71 @@
+    if (!node.chainman) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Node chainman not found");
+    }
+    WITH_LOCK(::cs_main, CHECK_NONFATAL(std::addressof(g_chainman) == std::addressof(*node.chainman)));
```
This causes a link failure for libbitcoin_common.

Perhaps move these `Ensure*` functions to a new `src/rpc/server_util.cpp` which is part of libbitcoin_server. `src/rpc/util` is for utility functions that are shared between server and client.

Sjors replied:
I see. For some reason macOS is totally happy with it, but on Linux I can see it blow up.

Sjors commented:
I added `server_util.cpp`. This seems to make the linker happy.

prusnak commented on src/Makefile.am:
```diff
@@ -199,12 +199,12 @@ BITCOIN_CORE_H = \
   rpc/register.h \
   rpc/request.h \
   rpc/server.h \
+	rpc/server_util.h \
```
use spaces, not tab

prusnak commented on src/Makefile.am:
```diff
@@ -345,6 +345,7 @@ libbitcoin_server_a_SOURCES = \
   rpc/net.cpp \
   rpc/rawtransaction.cpp \
   rpc/server.cpp \
+	rpc/server_util.cpp \
```
use spaces, not tab

prusnak commented on src/Makefile.am:
```diff
@@ -534,7 +535,7 @@ libbitcoin_common_a_SOURCES = \
   rpc/rawtransaction_util.cpp \
   rpc/external_signer.cpp \
-  rpc/util.cpp \
+	rpc/util.cpp \
```
use spaces, not tab

MarcoFalke commented on src/rpc/server_util.h:
```diff
@@ -0,0 +1,30 @@
+
+/**
+ * Ensure* helpers
+ */
```
Can remove this comment? The code is already self-documenting and one can see that all helpers in this section start with `Ensure`.

Sjors commented:
Comments addressed.

jnewbery commented:
The includes could be improved to include what's needed and not include what's not needed:

```diff
diff --git a/src/rpc/server_util.cpp b/src/rpc/server_util.cpp
index dfe07f5c3d..046a630b51 100644
--- a/src/rpc/server_util.cpp
+++ b/src/rpc/server_util.cpp
@@ -2,12 +2,18 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#include <rpc/server_util.h>
+
 #include <net_processing.h>
 #include <node/context.h>
-#include <rpc/server_util.h>
+#include <policy/fees.h>
+#include <rpc/protocol.h>
+#include <rpc/request.h>
+#include <txmempool.h>
 #include <util/system.h>
 #include <validation.h>
-#include <validationinterface.h>
+
+#include <any>
 
 NodeContext& EnsureAnyNodeContext(const std::any& context)
 {
diff --git a/src/rpc/server_util.h b/src/rpc/server_util.h
index c330efd0e6..dd8e8a6042 100644
--- a/src/rpc/server_util.h
+++ b/src/rpc/server_util.h
@@ -5,11 +5,11 @@
 #ifndef BITCOIN_RPC_SERVER_UTIL_H
 #define BITCOIN_RPC_SERVER_UTIL_H
 
-#include <rpc/protocol.h>
-#include <rpc/request.h>
+#include <any>
 
 class CBlockPolicyEstimator;
 class CConnman;
+class ChainstateManager;
 class CTxMemPool;
 struct NodeContext;
 class PeerManager;
```

Sjors commented:
Also done.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,39 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
            (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, m_chainparams.GetConsensus()) < STALE_RELAY_AGE_LIMIT);
 }
 
+bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex* index)
```
If you're taking a pointer, you should check that it's non-null before dereferencing it. Even better, pass be reference to communicate that this *must* be called with a valid `CBlockIndex`:

<details>
<summary>Diff</summary>

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index a822960dc2..f552aa36b5 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -246,7 +246,7 @@ public:
 
     /** Implement PeerManager */
     void CheckForStaleTipAndEvictPeers() override;
-    bool FetchBlock(NodeId id, const CBlockIndex* index) override;
+    bool FetchBlock(NodeId id, const CBlockIndex& index) override;
     bool GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const override;
     bool IgnoresIncomingTxs() override { return m_ignore_incoming_txs; }
     void SendPings() override;
@@ -1253,7 +1253,7 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
            (GetBlockProofEquivalentTime(*pindexBestHeader, *pindex, *pindexBestHeader, m_chainparams.GetConsensus()) < STALE_RELAY_AGE_LIMIT);
 }
 
-bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex* index)
+bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex& index)
 {
     PeerRef peer = GetPeerRef(id);
     if (peer == nullptr) return false;
@@ -1263,12 +1263,12 @@ bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex* index)
     if (state == nullptr) return false;
     if (!state->fHaveWitness) return false;
     const int node_sync_height = state->pindexBestKnownBlock ? state->pindexBestKnownBlock->nHeight : -1;
-    if (index->nHeight > node_sync_height) {
+    if (index.nHeight > node_sync_height) {
         return false;
     }
-    std::vector<CInv> invs{CInv(MSG_BLOCK | MSG_WITNESS_FLAG, index->GetBlockHash())};
+    std::vector<CInv> invs{CInv(MSG_BLOCK | MSG_WITNESS_FLAG, index.GetBlockHash())};
     // Mark block as in-flight unless it already is
-    if (!MarkBlockAsInFlight(id, index->GetBlockHash(), index)) return false;
+    if (!MarkBlockAsInFlight(id, index.GetBlockHash(), &index)) return false;
     bool success = m_connman.ForNode(id, [this, invs](CNode* pnode) {
         const CNetMsgMaker msgMaker(pnode->GetCommonVersion());
         this->m_connman.PushMessage(pnode, msgMaker.Make(NetMsgType::GETDATA, invs));
@@ -1276,12 +1276,12 @@ bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex* index)
     });
     if (success) {
         LogPrint(BCLog::NET, "Requesting block %s from peer=%d\n",
-                index->GetBlockHash().ToString(), id);
+                index.GetBlockHash().ToString(), id);
     } else {
         // Remove block from in-flight
-        MarkBlockAsReceived(index->GetBlockHash());
+        MarkBlockAsReceived(index.GetBlockHash());
         LogPrint(BCLog::NET, "Failed to request block %s from peer=%d\n",
-                index->GetBlockHash().ToString(), id);
+                index.GetBlockHash().ToString(), id);
     }
     return success;
 }
diff --git a/src/net_processing.h b/src/net_processing.h
index 577c925562..659504397f 100644
--- a/src/net_processing.h
+++ b/src/net_processing.h
@@ -43,7 +43,7 @@ public:
     virtual ~PeerManager() { }
 
     /** Attempt to manually fetch block from a given node. */
-    virtual bool FetchBlock(const NodeId nodeid, const CBlockIndex* pindex) = 0;
+    virtual bool FetchBlock(const NodeId nodeid, const CBlockIndex& pindex) = 0;
 
     /** Get statistics from node state */
     virtual bool GetNodeStateStats(NodeId nodeid, CNodeStateStats& stats) const = 0;
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index aa6b216ab1..cb3f2f3abc 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -746,7 +746,7 @@ static RPCHelpMan getblockfrompeer()
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Block not found");
     }
 
-    if (!peerman.FetchBlock(nodeid, pblockindex)) {
+    if (!peerman.FetchBlock(nodeid, *pblockindex)) {
         throw JSONRPCError(RPC_MISC_ERROR, "Failed to fetch block from peer");
     }
     return UniValue::VOBJ;
```

</details>

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,39 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
 }
 
+bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex* index)
+{
```
Consider adding some line breaks to this function to split up the units of logic. It's a bit of a wall of text right now.

Sjors replied:
I added some line breaks and prose.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
 
+bool PeerManagerImpl::FetchBlock(NodeId id, const CBlockIndex& index)
+{
+    PeerRef peer = GetPeerRef(id);
+    if (peer == nullptr) return false;
```
`peer` isn't used. You can remove it.

Sjors replied:
This is the only place where we check that this peer actually exists (i.e. wasn't disconnected). I'll add a comment.

jnewbery replied:
Actually, you check below that a `CNodeState` exists, which can only happen if the peer isn't disconnected, which makes this check redundant.

Sjors replied:
Ok, I'll remove it then.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+{
+    PeerRef peer = GetPeerRef(id);
+    if (peer == nullptr) return false;
+    if (fImporting && !fReindex) return false;
```
Why not match the logic in `ProcessMessage()` when receiving a block:

```suggestion
    if (fImporting || fReindex) return false;
```

Sjors replied:
I think I copied this one from another method. But fewer negations make sense.

jnewbery commented on src/net_processing.h:
```diff
@@ -42,6 +42,9 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
     virtual ~PeerManager() { }
 
+    /** Attempt to manually fetch block from a given node. */
+    virtual bool FetchBlock(const NodeId nodeid, const CBlockIndex& pindex) = 0;
```
Remove `const` from the pass-by-value argument:

```suggestion
    virtual bool FetchBlock(NodeId nodeid, const CBlockIndex& pindex) = 0;
```

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (!MarkBlockAsInFlight(id, index.GetBlockHash(), &index)) return false;
+
+    // Send block request message to the peer
+    bool success = m_connman.ForNode(id, [this, invs](CNode* pnode) {
```
Prefer to capture `invs` by reference than making a copy:

```suggestion
    bool success = m_connman.ForNode(id, [this, &invs](CNode* pnode) {
```

Also maybe rename pnode to node if you're touching this part again.

jnewbery commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,58 @@
+    assert_raises_rpc_error,
+)
+
+import time
```
Import standard library modules first, then local modules.

jnewbery commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,58 @@
+
+    def run_test(self):
+        self.log.info("Mine 4 blocks on Node 0")
+        self.nodes[0].generatetoaddress(4, self.nodes[0].get_deterministic_priv_key().address)
```
```suggestion
        self.nodes[0].generate(4)
```

Same below.

jnewbery commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,58 @@
+
+        self.log.info("Connect nodes to sync headers")
+        self.connect_nodes(0, 1)
+        self.sync_blocks(self.nodes[0:2])
```
```suggestion
        self.sync_blocks()
```

jnewbery commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,58 @@
+        assert_equal(len(peers), 1)
+        peer_0_peer_1_id = peers[0]["id"]
+        self.nodes[0].getblockfrompeer(short_tip, peer_0_peer_1_id)
+        time.sleep(1)
+        self.nodes[0].getblock(short_tip)
```
Perhaps poll with `wait_until` rather than a sleep (sleep will take needlessly long on a fast system and may time out on a slow system).

Sjors replied:
Done, though it's rather involved afaik there's no one-liner to say "True if x doesn't throw"

jnewbery replied:
Change looks good. Thanks!

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+
+    if (success) {
+        LogPrint(BCLog::NET, "Requesting block %s from peer=%d\n",
+                index.GetBlockHash().ToString(), id);
```
```suggestion
                 index.GetBlockHash().ToString(), id);
```

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1252,6 +1253,47 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+        // Remove block from in-flight
+        MarkBlockAsReceived(index.GetBlockHash());
+        LogPrint(BCLog::NET, "Failed to request block %s from peer=%d\n",
+                index.GetBlockHash().ToString(), id);
```
```suggestion
                 index.GetBlockHash().ToString(), id);
```

prayank23 commented:
> I think it would be better to just do the right thing if you `getblock` a block you don't have... Don't make users figure out which peer has it...?

I agree with @luke-jr 

Or this RPC can make sense if all peers would reject such request by _default_ but if someone is okay with it, can save something like `allowgetblockfrompeer=1` in bitcoin.conf

Sjors commented:
It's not that simple. Do we fetch the block from all peers at once? That's a huge bandwidth waste if you have 100 connection. Do we ask them one by one in sequence? What's the timeout? Do we do in parallel? Hence my suggestion to leave that to a followup.

jnewbery commented on src/rpc/blockchain.cpp:
```diff
@@ -792,6 +745,43 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getblockfrompeer()
+{
+    return RPCHelpMan{"getblockfrompeer",
+                "\nAttempt to fetch block from a given peer.\n",
```
Perhaps document that this node must already have the block header in order to fetch the block (or relax that requirement?)

Sjors replied:
By "this node" you mean the node itself? What happens when you fetch a block that you don't have the header for?

Sjors replied:
I see `MarkBlockAsInFlight` can handle a `nullprtr` index. It might indeed be simpler to drop this requirements, as well as the `index.nHeight > node_sync_height` check.

jnewbery replied:
Sorry, yes I meant "we" as "this node" (the node that you're running this command on)

Sjors commented:
Rebased, dropped requirement for local node to have the header, dropped redundant `GetPeerRef`, dropped requirement for other node to be at the right height.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -247,6 +247,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Implement PeerManager */
     void CheckForStaleTipAndEvictPeers() override;
+    bool FetchBlock(NodeId id, uint256 hash, const CBlockIndex* index) override;
```
Consider passing `uint256` by const reference (to avoid unnecessary copying, generally only basic types should be passed by value).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -247,6 +247,7 @@ class PeerManagerImpl final : public PeerManager
 
     /** Implement PeerManager */
     void CheckForStaleTipAndEvictPeers() override;
+    bool FetchBlock(NodeId id, uint256 hash, const CBlockIndex* index) override;
```
Maybe document that `index` may be null here?

Sjors replied:
Added function documentation.

jnewbery reviewed (commented):
utACK 803d2f568f7f24557543f61bf9d5a73952caedb5.

Changes look safe.

Sjors committed:
rpc: getblockfrompeer

Co-authored-by: John Newbery <john@johnnewbery.com>

sipa commented:
Is it really necessary to make this specify which peer? It doesn't sound too hard to make the existing block fetching logic support extra requested blocks, I think.

Sjors commented:
@sipa the current block fetching logic is buried deep inside net_processing. A `MSG_BLOCK` is generated when processing received headers:

https://github.com/bitcoin/bitcoin/blob/4741aec1dd28829f45abcc529cddaa0ff04d07a0/src/net_processing.cpp#L2011-L2021

Or when receiving a block:
https://github.com/bitcoin/bitcoin/blob/4741aec1dd28829f45abcc529cddaa0ff04d07a0/src/net_processing.cpp#L1733-L1739

There doesn't seem to be a queue of blocks to fetch that's distributed over peers and checked in a background process.

sipa commented:
@Sjors Sure, but that doesn't sound too hard to add. And doing it that way sounds a lot more useful.

Sjors commented:
In practice I'm already able to use this PR to fetch stale blocks by simply looping over my peers and "spamming" them all. I then disconnect if the block doesn't show up, connect to random new peers and then try again.

So although the current incarnation is not ideal, it's already useful. A revamp of `ProcessHeadersMessage` with a separate block download manager sounds doesn't sound easy to me. It also needs different timeout rules than IBD, because most of the time a peer won't have the block.

The use case of re-fetching pruned blocks is more similar to IBD, since you can expect every non-pruned node  with sufficiently high `pindexLastCommonBlock` to have it.

sipa commented:
Yes, I imagine it'd use the "background' fetching logic in SendMessages, not the direct fetch in response to an announcement one.

Sjors commented:
@sipa we don't fetch a block if we already have one for that height. So if you're trying retrieve the losing block from a race, most likely your peers also saw it later and didn't fetch it.

sipa commented:
Oh, I see. I missed that fetching non-main-chain blocks was a goal here.

Why would you want that?

Sjors commented:
That's indeed the main reason I wrote this, though I can see how it's useful for redownloading pruned blocks too.

I'm using this for ForkMonitor to fetch stale blocks and compare the transactions, to see if there was a double-spend: https://forkmonitor.info/stale/btc/679823

MarcoFalke commented:
Error: RPC command "getblockfrompeer" not found in RPC_COMMANDS_SAFE_FOR_FUZZING or RPC_COMMANDS_NOT_SAFE_FOR_FUZZING. Please update test/fuzz/rpc.cpp.

Sjors commented:
Rebased and marked safe for fuzzing (but I haven't tried that).

jnewbery commented:
ACK 1df0fd9aceaeaf8b45cb0474ce44135418c7d5b7

Sjors commented:
Rebased due to silent merge conflict with #22141, which I randomly caught while looking at IRC notifications :-) But it's a welcome simplification.

Sjors commented:
Rebased after some `Ensure*` methods were renamed in #21391.

In addition since #22221, `BlockRequested` no longer takes a pointer. Which means we can't call it for blocks for which we don't have a header. Paging @MarcoFalke to check if that was indeed the intention.

MarcoFalke commented:
See also #10794

Sjors commented:
@MarcoFalke specifically its `requestblock` RPC? There's been too much refactoring to simply cherry-pick it, but it could contain inspiration for improvement.

MarcoFalke commented:
Yes, just left the comment to remind myself that this has been worked on in the past already. No need to change anything here, but if someone wants to read some related threads, it might be helpful.

jnewbery reviewed (commented):
> In addition since #22221, BlockRequested no longer takes a pointer. Which means we can't call it for blocks for which we don't have a header. 

What do you think about restricting this RPC to where the node already has the header, by returning in `getblockfrompeer()` if `LookupBlockIndex()` fails? That'd make the functionality in net_processing more closely mirror our normal block download, where we'll only ever request a block if we already have its header. If the node doesn't already have the header, the user could first call `submitheader` with the serialized block header, followed by `getblockfrompeer()`.

LarryRuane commented on src/rpc/blockchain.cpp:
```diff
@@ -747,6 +749,39 @@ static RPCHelpMan getmempoolentry()
+    if (!peerman.FetchBlock(nodeid, hash, pblockindex)) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Failed to fetch block from peer");
+    }
+    return UniValue::VOBJ;
```
Why do we not return `NullUniValue` here? Returning `UniValue::VOBJ` causes `bitcoin-cli` to print empty JSON object; returning `NullUniValue` would print nothing (as many other RPCs do). Is it to make this more backward-compatible if we later make it return something, like the block?

Sjors replied:
The first version of this PR included a `success` boolean, but that didn't make sense because we don't know if we actually get the block. So I guess I made the response an empty VOBJ. I'll switch to Null.

LarryRuane commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,62 @@
+        self.sync_blocks()
+
+        self.log.info("Node 0 should only have the header for node 1's block 3")
+        short_tip_synced = False
+        for x in self.nodes[0].getchaintips():
+            if x['hash'] == short_tip:
+                assert_equal(x['status'], "headers-only")
+                short_tip_synced = True
+        assert short_tip_synced
```
```suggestion
        for x in self.nodes[0].getchaintips():
            if x['hash'] == short_tip:
                assert_equal(x['status'], "headers-only")
                break
        else:
            raise AssertionError("short tip not synced")
```

jnewbery replied:
Much more pythonic :+1:

LarryRuane commented on test/functional/rpc_getblockfrompeer.py:
```diff
@@ -0,0 +1,62 @@
+        assert_equal(len(peers), 1)
+        peer_0_peer_1_id = peers[0]["id"]
+        self.nodes[0].getblockfrompeer(short_tip, peer_0_peer_1_id)
+        self.wait_until(lambda: self.check_for_block(short_tip), timeout=1)
```
Suggestions for additional tests:
```
        self.log.info("No error if block is not found in the index")
        assert_equal(self.nodes[0].getblockfrompeer("12" * 32, peer_0_peer_1_id), None)
        self.log.info("Non-existent peer generates error")
        assert_raises_rpc_error(-1, "Failed to fetch block from peer", self.nodes[0].getblockfrompeer, short_tip, peer_0_peer_1_id + 1)
```

Sjors replied:
Thanks, added the second one.

LarryRuane reviewed (commented):
ACK 30e2ba6c4cef9248c53542faeb8ec768079fc5fb
Suggestions optional

mzumsande commented:
> What do you think about restricting this RPC to where the node already has the header, by returning in `getblockfrompeer()` if `LookupBlockIndex()` fails?

I agree that this would make sense. My understanding is that if we'd request a block for which we don't  have the header and which has less work than our tip, then `getblockfrompeer` would currently signal success, we might also get the block delivered by the peer, but then `AcceptBlock()` would never store it because `fRequested` and `fHasMoreOrSameWork` would be false ([code](https://github.com/bitcoin/bitcoin/blob/e614cc8cd808902c843568f1790730622d619620/src/validation.cpp#L3421-L3431 ))
But if we cannot accept the block, I don't think it makes sense to request it in the first place (and we should instead report an RPC error).

Sjors commented:
I reverted back to returning an empty object as per the [developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#rpc-interface-guidelines) "Try to make the RPC response a JSON object".

Also made `FetchBlock` take `CBlockIndex` as a reference rather than a pointer, since it's now mandatory.

jnewbery commented:
reACK 29263ae14a81d4cc55de61f51db981edf54399d0

laanwj commented:
Concept ACK

0xB10C commented on src/rpc/blockchain.cpp:
```diff
@@ -18,6 +18,8 @@
 #include <hash.h>
 #include <index/blockfilterindex.h>
 #include <index/coinstatsindex.h>
+#include <net.h> // For NodeId
```
nit: not sure if there is need for a `// For NodeId` comment here. I think comments like these tend to be overlooked when adding something else that needs `net.h` and thus out-dating the comment.

fanquake replied:
Yea, please don't add comments like these.

0xB10C commented on src/rpc/blockchain.cpp:
```diff
@@ -748,6 +750,44 @@ static RPCHelpMan getmempoolentry()
+    if (!index) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Block header missing");
+    }
+
```
optional but maybe worthwile: could check if the block is already available and return true if it is. Could reduce bandwidth usage in cases where users spam/iterate their all their peers with a `getblockfrompeer`.

Sjors replied:
The result now includes a warning if we already had the block. This seems better than throwing an error, since an RPC client might attempt fetching for multiple peers in parallel.

0xB10C reviewed (commented):
I've manually tested `getblockfrompeer` against two peers on a custom signet. Both peers share the chain tip, but both know about different (now stale) branches. I purposefully black-box tested the RPC without look at the code and tests added in this PR.

Tests:

1. Requesting a block without having the header: returns the error message `Block header missing`. This behavior is documented and expected :heavy_check_mark:.  
2. Requesting a block from a peer that does not know about the block: Returns with `{}` and a `getblock` RPC call returns with `Block not found on disk`. This behavior is as expected as peer can't send the block if it does not know about it :heavy_check_mark:.
3. Requesting a block from a peer that hasn't validated the block (`getchaintips` status is `valid-headers` which is documented as ` All blocks are available for this branch, but they were never fully validated`): Returns with `{}` and a `getblock` RPC call returns with `Block not found on disk`. This behavior is fine I guess. In normal operation, we probably don't want to serve unvalidated blocks to peers as a DoS mitigation. The RPC likely inherits this behavior :question:.
4. Requesting a block from a peer that has only the block header (`getchaintips` status is 'headers-only'):  Returns with `{}` and a `getblock` RPC call returns with `Block not found on disk`. This behavior is as expected as peer can't send the block if it only knows about the header :heavy_check_mark:.
5. Requesting a block from a peer that has validated the block but it's not in the active chain (`getchaintips` status is `valid-fork`):  Returns with `{}` and a `getblock` RPC call returns the block :heavy_check_mark:.
6. Requesting a block from a peer that has marked as 'invalid':  Returns with `{}` and a `getblock` RPC call fails. The peer does not send the block likely to a similar reason as in test 3 :question:.


Notes:

Requesting an unavailable block from a peer `A` multiple times in a row fails with the error message `Failed to fetch block from peer` starting with the second request. A `getdata` message is only send to the peer on the first attempt. The error message is not displayed when requesting from a peer `B` in between the requests to peer `A`.

Re-requesting already available blocks resends a `getdata` message and the peer responds with the block again. Maybe check if a block is available before requesting it?

My node disconnected from a peer at some point because a `getdata` timed out. `Timeout downloading block 00000268... from peer=0, disconnecting`. This block was in a branch marked as `valid-headers` by peer 0. My node directly reconnected to peer 0 as peer 2 as it's manually configured with `addnode`.  **This could be an issue when using this in the real world.**

Requesting blocks that would be an orphan to us (the previous header is unknown) is not possible as `submitheader` requires the previous header to be known. I'm not sure there is another way to have orphan (orphan, not stale) blocks or headers in the block tree.


---

Some code review comments below too. 

---

RE my own comment:
> Requesting an unavailable block from a peer `A` multiple times in a row fails with the error message `Failed to fetch block from peer` starting with the second request. A `getdata` message is only send to the peer on the first attempt. The error message is not displayed when requesting from a peer `B` in between the requests to peer `A`.

Looked this behavior up in the code: `PeerManagerImpl::BlockRequested()` returns false if the block is already requested. This causes the RPC to fail too. If the block is requested from another peer, the request is dropped. 

https://github.com/bitcoin/bitcoin/blob/29263ae14a81d4cc55de61f51db981edf54399d0/src/net_processing.cpp#L867-L868

Doesn't this cause us to drop a request to peer A if A is slow and we instantly request it from B too? Could it cause issues if if A then sends us the block or do we just accept it?  

Generally seems like doing `getblockfrompeer` for many peers without a delay between peers is not the ideal use of this RPC.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -748,6 +750,44 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getblockfrompeer()
+{
+    return RPCHelpMan{"getblockfrompeer",
+                "\nAttempt to fetch block from a given peer.\n"
```
I added a sentence to the RPC doc.

Sjors commented:
Thanks for the thorough testing @0xB10C.

Regarding (3) and (6): if a peer doesn't give us the block, there's nothing we can do. Since the call is not asynchronous, we can't give the user much useful feedback either. Changing p2p behavior for peers to give us a block before it's evaluated, or one it considers invalid, is beyond the scope of this PR.

> Requesting an unavailable block from a peer `A` multiple times in a row fails with the error message `Failed to fetch block from peer` starting with the second request. A `getdata` message is only send to the peer on the first attempt.

Interesting. I'd rather not make the RPC aware of too much internal magic of netmanager. If the user calls an RPC again after an error message, I guess they're on their own :-)

> The error message is not displayed when requesting from a peer `B` in between the requests to peer `A`.

The only reason I can think of for requesting a block from multiple peers, is that the user is happy as long as at least one succeeds. So this is probably not a huge deal, but still strange. I could add a recommendation to the help that it's safer to call this method in sequence with a few second delay?

Sjors commented:
Rebased after #22650 (trivial conflict with first commit).

jnewbery commented:
reACK 4096e6fb013aa3fbe3463d24b70310ad8517da5f

jnewbery commented:
trivial reACK 9181e2e217

Confirmed with `git rangediff` that the only conflict was for touching lines.

LarryRuane commented on src/net_processing.cpp:
```diff
@@ -1426,6 +1427,41 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (!state->fHaveWitness) return false;
+
+    // Construct message to request the block
+    std::vector<CInv> invs{CInv(MSG_BLOCK | MSG_WITNESS_FLAG, hash)};
```
nit, if you have a chance to retouch, could this be moved down to just before it's first used by the `m_connman.ForNode()` call below?

Sjors replied:
I moved it inside the `ForNode()` block

LarryRuane reviewed (commented):
ACK 9181e2e217955fff5f726b0e011afb78fb90af8f

jamesob commented:
Concept ACK, will review in the next few days.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -796,6 +750,52 @@ static RPCHelpMan getmempoolentry()
+    UniValue result = UniValue::VOBJ;
+
+    if (index->nStatus & BLOCK_HAVE_DATA) {
+        result.pushKV("warnings", "Block already downloaded");
```
Is it intentional that this succeeds even with an invalid nodeid?

Perhaps the tests should check for the expected behaviour.

Sjors replied:
It's probably a good sanity check to add. I could call `GetNodeStateStats` on the peer as a way to find out if it's valid, though that's not strictly what that PeerMan function is for. @jnewbery

jnewbery replied:
You could use `CConnman::ForNode()` to determine whether the peer with that nodeid exists. Here's the diff (with the required changes to make the tests run on master):

```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 9fc586c122..fdbfe55aaf 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -780,11 +780,17 @@ static RPCHelpMan getblockfrompeer()
     const NodeContext& node = EnsureAnyNodeContext(request.context);
     ChainstateManager& chainman = EnsureChainman(node);
     PeerManager& peerman = EnsurePeerman(node);
+    CConnman& connman = EnsureConnman(node);
 
     uint256 hash(ParseHashV(request.params[0], "hash"));
 
     const NodeId nodeid = static_cast<NodeId>(request.params[1].get_int64());
 
+    // Check that the peer with nodeid exists
+    if (!connman.ForNode(nodeid, [](CNode* node) {return true;})) {
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("Peer nodeid %d does not exist", nodeid));
+    }
+
     const CBlockIndex* const index = WITH_LOCK(cs_main, return chainman.m_blockman.LookupBlockIndex(hash););
 
     if (!index) {
diff --git a/test/functional/rpc_getblockfrompeer.py b/test/functional/rpc_getblockfrompeer.py
index ec562c94f9..e841cba70e 100755
--- a/test/functional/rpc_getblockfrompeer.py
+++ b/test/functional/rpc_getblockfrompeer.py
@@ -27,11 +27,11 @@ class GetBlockFromPeerTest(BitcoinTestFramework):
 
     def run_test(self):
         self.log.info("Mine 4 blocks on Node 0")
-        self.nodes[0].generate(4)
+        self.generate(self.nodes[0], 4, sync_fun=self.no_op)
         assert_equal(self.nodes[0].getblockcount(), 204)
 
         self.log.info("Mine competing 3 blocks on Node 1")
-        self.nodes[1].generate(3)
+        self.generate(self.nodes[1], 3, sync_fun=self.no_op)
         assert_equal(self.nodes[1].getblockcount(), 203)
         short_tip = self.nodes[1].getbestblockhash()
 
@@ -60,7 +60,7 @@ class GetBlockFromPeerTest(BitcoinTestFramework):
         assert_raises_rpc_error(-1, "Block header missing", self.nodes[0].getblockfrompeer, "00" * 32, 0)
 
         self.log.info("Non-existent peer generates error")
-        assert_raises_rpc_error(-1, "Failed to fetch block from peer", self.nodes[0].getblockfrompeer, short_tip, peer_0_peer_1_id + 1)
+        assert_raises_rpc_error(-1, f"Peer nodeid {peer_0_peer_1_id + 1} does not exist", self.nodes[0].getblockfrompeer, short_tip, peer_0_peer_1_id + 1)
 
         self.log.info("Successful fetch")
         result = self.nodes[0].getblockfrompeer(short_tip, peer_0_peer_1_id)
```

Sjors replied:
Done

jnewbery replied:
This branch currently has a silent merge conflict with master. You'll need to rebase and take the `sync_fun=self.no_op` bit before merge.

Sjors replied:
Done

fjahr reviewed (commented):
tested ACK 9181e2e217955fff5f726b0e011afb78fb90af8f

I think this is ready to be merged as-is and my comments below can be addressed in a follow-up.

The documentation should be improved on how long users can expect this block to be around on a pruned node. If someone uses this on a `prune=550` node, for example, this may become confusing for certain use cases. I made a suggestion in my comment but I can also propose this in a follow-up PR since it might require further discussion.

I also wrote an additional test for the pruned node use case [here](https://github.com/fjahr/bitcoin/commit/32efa3017a092ae78782e9e3091e2f2499e1c1dc). Feel free to add it here if you have to retouch, otherwise, I will open a follow-up PR with it.

luke-jr commented on src/rpc/blockchain.cpp:
```diff
@@ -748,6 +750,52 @@ static RPCHelpMan getmempoolentry()
+    return RPCHelpMan{"getblockfrompeer",
+                "\nAttempt to fetch block from a given peer.\n"
+                "\nWe must have the header for this block, e.g. using submitheader.\n"
+                "\nReturns {} if a block-request was successfully scheduled\n",
```
Even that is not guaranteed at the moment (albeit very likely to hold in practice).

Combined with #19463, the caller could set a prune lock to ensure it doesn't get re-pruned before the caller is done with it.

fjahr replied:
So far I wasn't able to figure out in which cases this could not be guaranteed. I opened a follow-up PR in #23813. Please let me know if you have suggestions on improving the text there. Sure #19463 can be helpful but my aim with this for now is to just make sure that users don't get wrong expectations, like that the block would be available forever for example.

jnewbery commented:
ACK dce8c4c38111556ca480aa0e63c46b71f66b508f

fjahr commented:
re-ACK dce8c4c38111556ca480aa0e63c46b71f66b508f

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1427,6 +1428,41 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+    if (!state->fHaveWitness) return false;
+
+    // Mark block as in-flight unless it already is
+    if (!BlockRequested(id, index)) return false;
```
maybe add a comment that this will re-assign the block if it is in flight from another peer?

Sjors replied:
I'll also update the RPC doc, because it looks like `BLOCKTXN` are ignored entirely if they're no longer on the in flight list for a given peer.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1427,6 +1428,41 @@ bool PeerManagerImpl::BlockRequestAllowed(const CBlockIndex* pindex)
+                 hash.ToString(), id);
+    } else {
+        RemoveBlockRequest(hash);
+        LogPrint(BCLog::NET, "Failed to request block %s from peer=%d\n",
```
Seems odd to log a failure to the debug log on an rpc call. I'd expect the error to go to the user instead. Also, shouldn't this say "peer not found"?

Sjors replied:
We already check that the peer exists above, so the failure would be "not fully connected"; changed...

MarcoFalke commented on src/net_processing.h:
```diff
@@ -42,6 +42,16 @@ class PeerManager : public CValidationInterface, public NetEventsInterface
+     * Attempt to manually fetch block from a given peer. We must already have the header.
+     *
+     * @param[in]  id       The peer id
+     * @param[in]  hash     The block hash
+     * @param[in]  pindex   The blockindex
```
why are you passing the hash twice? the blockindex already holds its own hash, no?

Also, `pindex` is not a pointer, so should be called `block_index`.

Sjors replied:
That may have gotten mixed up over the course of rebases and changes, I'll take a second look.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -803,6 +757,58 @@ static RPCHelpMan getmempoolentry()
+                },
+                RPCResult{RPCResult::Type::OBJ, "", "",
+                {
+                    {RPCResult::Type::STR, "warnings", "any warnings"}
```
missing optional

Sjors replied:
Thanks for adding in #23702

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -803,6 +757,58 @@ static RPCHelpMan getmempoolentry()
+    PeerManager& peerman = EnsurePeerman(node);
+    CConnman& connman = EnsureConnman(node);
+
+    uint256 hash(ParseHashV(request.params[0], "hash"));
```
nit: can be const

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -803,6 +757,58 @@ static RPCHelpMan getmempoolentry()
+
+    uint256 hash(ParseHashV(request.params[0], "hash"));
+
+    const NodeId nodeid = static_cast<NodeId>(request.params[1].get_int64());
```
any reason to use a narrowing cast and not the safe alternative?

```cpp
     const NodeId nodeid{request.params[1].get_int64()};

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -803,6 +757,58 @@ static RPCHelpMan getmempoolentry()
+
+    // Check that the peer with nodeid exists
+    if (!connman.ForNode(nodeid, [](CNode* node) {return true;})) {
+        throw JSONRPCError(RPC_MISC_ERROR, strprintf("Peer nodeid %d does not exist", nodeid));
```
Since the FetchBlock call has to deal with this anyway, might be better to remove this code and let FetchBlock deal with it.

MarcoFalke commented on src/rpc/blockchain.cpp:
```diff
@@ -803,6 +757,58 @@ static RPCHelpMan getmempoolentry()
+static RPCHelpMan getblockfrompeer()
+{
+    return RPCHelpMan{"getblockfrompeer",
+                "\nAttempt to fetch block from a given peer.\n"
```
Should also mention that this will disconnect the peer if the peer pretends to not have the block

Sjors replied:
That seems a bit too deep in the net_processing weeds?

Sjors commented:
@MarcoFalke thanks, working on a followup PR...

MarcoFalke commented:
could make sense to base on https://github.com/bitcoin/bitcoin/pull/23702 to avoid conflicts

Sjors commented:
Followups in #23706
